lib/libutils/ext/include/fault_mitigation.h

```diff
@@ -0,0 +1,619 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2022, Linaro Limited
+ */
+#ifndef __FAULT_MITIGATION_H
+#define __FAULT_MITIGATION_H
+
+#include <assert.h>
+#include <config.h>
+#include <string.h>
+#include <util.h>
+
+#ifdef __KERNEL__
+#include <kernel/panic.h>
+#include <kernel/thread.h>
+#else
+#include <tee_api.h>
+#endif
+
+/*
+ * Fault migitigation helpers to make successful Hardware Fault Attacks
+ * harder to achieve. The paper [1] by Riscure gives background to the
+ * problem.
+ *
+ * These helpers aim to make it hard for a single glitch attack to succeed
+ * while the protected function or one of the ftmn_*() functions are
+ * executed.
+ *
+ * To have something to work with we assume that a single glitch may affect
+ * a few instructions in sequence to do nothing or to corrupt the content
+ * of a few registers.
+ *
+ * [1] https://www.riscure.com/uploads/2020/05/Riscure_Whitepaper_Fault_Mitigation_Patterns_final.pdf
+ */
+
+#include <stdint.h>
+#include <stdbool.h>
+
+/*
+ * struct ftmn_check - track current checked state
+ * @steps:	accumulated checkpoints
+ * @res:	last stored result or return value
+ *
+ * While a function is executed it can update its state as a way of keeping
+ * track of important passages inside the function. Before the function
+ * returns with for instance ftmn_return_res() to check that the
+ * accumulated state matches the expected state.
+ *
+ * @res is xored with FTMN_DEFAULT_HASH in order to retrieve the saved
+ * result or return value.
+ */
+struct ftmn_check {
```

Comment:
This could in theory also be used across multiple static function. Anyway, I think "func" in name here would be redundant. So the question is do we prefer check or state? I prefer check as it's more specific, it records passed checkpoints.

Reply:
That's ok with me.